### PR TITLE
Add support for dynamic loading.

### LINKED ISSUE #28: OpenTracing support for dynamic loading.
Many C/C++ projects (e.g. nginx, haproxy, varnish, envoy) that would benefit from OpenTracing instrumentation are distributed as binary executables instead of libraries.

When instrumenting such projects, the practice so far has been to write code for each tracer that exposes its configuration options as part of the applications configuration. (See for example [nginx-opentracing/lightstep](https://github.com/opentracing-contrib/nginx-opentracing/tree/master/lightstep) and [nginx-opentracing/zipkin](https://github.com/opentracing-contrib/nginx-opentracing/tree/master/zipkin)).

This increases the cost of maintaining these projects and goes against OpenTracing's goal of making it easy to switch tracing implementations. Furthermore, since many of the projects are used as building blocks in larger systems that interact with the application's configuration, the cost compounds since those systems also need to be aware a tracer's configuration. (See for example how OpenTracing was incorporated into Kubernetes' [ingress-nginx controller](https://github.com/kubernetes/ingress-nginx/blob/master/rootfs/etc/nginx/template/nginx.tmpl#L18)).

I propose we adopt some conventions for C++ tracers that would allow applications to load tracers in dynamically using [dlopen](http://man7.org/linux/man-pages/man3/dlopen.3.html) and decouple applications from specific tracer configuration options.

This could be done by standardizing on a function that could be looked up from a tracer's shared library using [dlsym](https://linux.die.net/man/3/dlsym). Consider for example a C function (to avoid mangling) like
```
extern "C" int opentracing_make_vendor_tracer(void** tracer, const char* json_configuration);
```
that would take in a configuration string for the tracer in JSON, set `tracer`, and return 0 on success or some non-zero error code on failure.

A user of an application like nginx, then, might configure a tracer by adding something like
```
opentracing_tracer_option access_token abc123;
opentracing_tracer_option collector_host localhost;
opentracing_load_tracer liblightstep_tracer.so;
``` 
into `nginx.conf` which would cause nginx to lookup `opentracing_make_vendor_tracer`in `liblightstep_tracer.so` and call it with
```json
{
  "access_token": "abc123",
  "collector_host": "localhost"
}
```
avoiding the need to have specific configuration directives for each LightStep option.

Projects that want to avoid the pains associated with dynamic library management could still link in some tracers statically, but might also add dynamic loading functionality like this as an optional way to plug in additional tracers.

Standards around both trace encoding and transport (such as what's outlined [here](https://github.com/opentracing/specification/issues/86)) might make it possible some day to have a vendor neutral tracer that could report to any tracing system and avoid the need to dynamically load anything in, but this would address a more immediate need to make OpenTracing scalable and maintainable for these projects.

### PR DESCRIPTION
**Pull request by rnburn**
This PR adds support for dynamic loading to allow C++ applications instrumented with OpenTracing to dynamically load in tracing libraries at runtime instead of requiring them to be linked in at compile-time.

The basic idea is tracing libraries will implement the function `opentracing_make_tracer_factory` that constructs and returns a pointer to a `TracerFactory` which can be used construct the library's tracers from configuration strings. To dynamically load a tracing library, an application calls the function `dynamically_load_tracing_library ` which takes in the name of the library, calls the library's `opentracing_make_tracer_factory` function, and returns the `TracerFactory`.

Returning a factory like this instead of an actual tracer will allow for users to potentially reconfigure tracers by reconstructing a new replacement tracer from a different configuration string.

The `opentracing_make_tracer_factory` function is declared to have [weak linkage](https://en.wikipedia.org/wiki/Weak_symbol) so that multiple tracing libraries can be linked in without getting a multiple definition error. I know there was some concern about the portability of weak linkage, but its broadly supported by compilers on unix platforms and something similar can be done on windows with [`__declspec(selectany)`](https://msdn.microsoft.com/en-us/library/5tkz6s71.aspx). Still, I'm open to other approaches if anyone thinks there's a better way to solve this.

I currently have an example of dynamic loading working with a branch of [envoy](https://github.com/rnburn/envoy/tree/jaeger) and [jaeger](https://github.com/rnburn/cpp-client), which can be demoed with this docker [application](https://github.com/rnburn/jaeger-envoy-docker/tree/master/example). I also plan on adding a mock tracer library into opentracing-cpp which will support dynamic loading and can be used to add testing coverage for this new code. 

Fixes #28

### CODE REVIEW

**Comment by rnburn** [comment]
Something that's come up in the envoy review: https://github.com/envoyproxy/data-plane-api/pull/386#discussion_r159484483

Should we standardize on a tracer's configuration format being in json/yaml? Doing so would allow applications to specify the configuration inline.

cc @yurishkuro @isaachier

**Comment by yurishkuro** [comment]
@rnburn my 2 cents

"standadizing" is a difficult thing because each tracer's configuration is unique. If envoy needs a place in its own config that can be opaquely passed to the tracer, then it's envoy's business. E.g. it can say "tracerConfig" property is what contains the config, and stop "standardizing" at that point.

json/yaml again sounds like the concern of the containing application, not the tracer. In Jaeger clients we expose a Configuration object that *can be* initialized from a config file, but doesn't have to be. In fact, in Python/JS that Configuration simply takes a dict / object as constructor argument, which is idiomatic way of passing nested configuration. It's harder with compiled langs like Java/Go, where we expose an object/struct that contains say yaml annotations. But it would be better if there was some neutral configuration API where you can just say `cfg.getFloat("sampler.probability")`. Unfortunately, this is not something that's natively part of the language(s).

Finally, where standardizing does really make sense (aside from that cfg.getX API), is in passing that configuration to tracer factory, which is necessary for dynamic loading.

So in sumary
  * an API for reading config
  * a tracer factory API that takes the config reader
  * possible yet another factory API that takes static (tracer-level) tags if there is a need to standardize on those

**Comment by rnburn** [comment]
By standardize, I only meant to specify that the configuration is in json or yaml, not say anything about what's in the configuration.

Currently, the factory class for the tracer [takes](https://github.com/rnburn/opentracing-cpp/blob/dlopen/include/opentracing/tracer_factory.h#L49) a string representing the configuration. What form the string takes is the concern of both the application and the tracer since they have to agree on the format.

Ideally, in envoy's configuration you'd write something like

```yaml
...
tracerConfig:
    opt1: val11
    opt2:
        - val21
          val22
...
```

and envoy would call the factory method with the string
```yaml
opt1: val11
opt2:
    - val21
       val22
```
by agreeing that the tracer's configuration is in something like yaml, it can be easily embedded into the rest of envoy's configuration and at least someone validated (i.e. checked that it's valid yaml) before passing it to the tracer.

**Comment by yurishkuro** [comment]
> By standardize, I only meant to specify that the configuration is in json or yaml, not say anything about what's in the configuration.

I would split the concern of the config format from the config values. At minimum Tracer factory needs to have an API for a programmatically accessible config, such as `cfg.Get("sampler.priority")` . Then the actual format of the config (json, yaml, toml, xml, whatever) is no longer of concern for Tracer, but for the application that uses the config file.

**Comment by rnburn** [comment]
Couldn't the parsing of the configuration be handled by a library (e.g. [yaml-cpp](https://github.com/jbeder/yaml-cpp)) linked in by the tracer? It allows an accessible interface like what you're describing.

**Comment by yurishkuro** [comment]
My point is that I would make the parsing of the config separate from providing the config to the tracer. The latter is best done via some API. The parsing will typically be done by the enclosing application _anyway_, so why make it a tracer responsibility (with all side effects like extra dependencies, different formats, etc.)?

**Comment by rnburn** [comment]
But that would complicate the API since tracers would need to be constructed from some sort of object that allows it to query the configuration instead of a simple string.

Also, applications need not have to parse the tracer's configuration. If it's specified by a file, it can be read into a string and passed to the tracer library via a string.

**Comment by rnburn** [comment]
What about nested configurations or configurations with values other than string? Once you account for that the API gets a lot bigger than the simple key-value lookup you're proposing.

And I don't see linking in a json parsing library as being that burdensome -- it's likely something that tracer's would want to do anyways (Jaeger's already linking to [yaml-cpp](https://github.com/jaegertracing/cpp-client/blob/master/CMakeLists.txt#L83) btw).

**Comment by tedsuo** [comment]
@yurishkuro I'd prefer a string serialized into a standard data format for this API. A couple of reasons:

This API is separate from the tracer config API. The later is tracer-specific. It makes sense for that to be a non-serialized object which does not require JSON or other parsing. But this is a generalized transport that connects an application with any tracer, and any kind of option any tracer may want. 

Any non-standard API will create difficulty for application developers, as they will have to manually implement the interface and translate all of the tracer options from their former into it. That sounds like a lot of work that can't easily be wrapped up in a shared helper library. Limiting the API to less than JSON - only supporting non-nested `string` types for example - will increase that burden. Likewise on the tracer side of the fence - the tracer will have more work to do converting from a limited format into whatever format it wants. That problem could be compounded by conversions which have some ambiguity around their string representation. It would be much simpler to let the application re-serialize the tracer portion of the configuration file as JSON and pass it on.

Regarding the issue of tracers bundling in a JSON parser - the parser is only needed when the tracer is included as a shared object, not compiled in directly, so it should be possible to ship a version of your tracer that does not include a parsing library, and wrap it in a version that handles parsing and exposes a TracerFactory. If it turns out we can standardize on something more restrictive than JSON, we could write a small parsing library that tracers could include instead of a generalized JSON parser, and avoid bundling a larger dependency.

Export maps can be used on linux to only expose the Tracer API, and ensure there are no dependency conflicts.

**Comment by yurishkuro** [comment]
If this is only JSON string that's allowed, as opposed to a variety of formats, then I don't have a strong objection.

I am curious though what you all think about configuring via env vars, as advocated by https://12factor.net/

**Comment by tedsuo** [comment]
ENV vars were very problematic for us on Cloud Foundry (relevant since it’s so similar to Heroku). Primary issue is a single global namespace shared by multiple independent actors creates too much opportunity for collision. 

We ended up with “single env var containing a JSON string” for each actor to avoid this. So, basically, right back to this, only a more annoying invocation pattern for the application.

**Comment by yurishkuro** [comment]
@objectiser see interesting comment ^ re env vars

**Comment by objectiser** [comment]
Yes I guess it can be a problem, depending on the nature of the application and components/actors being configured. But in our context, where we have a single tracer per service (generally), it is a valid way to handle configuration.

**Comment by tedsuo** [comment]
I suppose there's nothing to stop a tracer from checking for ENV vars if it wants too... but I wouldn't recommend making it a requirement. ENV Vars are part of the execution context, you may be surprised at how hard it is to set them, for example in execution contexts where control of ENV vars are handed off to an actor that doesn't have the required tracer information.

It wold be better if dynamic loading included tracer configuration in a standard data format, with no ties to the execution environment, so that the application can control how it acquires the config data.

**Comment by isaachier** [comment]
Do you mind adding a CMake test for weak symbol support and then conditionally compiling the feature? That allows users on machines that do not support weak symbols to compile opentracing-cpp regardless. For example,

```
set(lib_src ...)  # List of library sources
try_compile(
  weak_symbol
  "${CMAKE_CURRENT_BINARY_DIR}/CMakeFiles/CMakeTmp"
  SOURCES "cmake/weak_symbol.cpp")
if(weak_symbols)
  list(APPEND lib_src ...)  # Append weak symbol support sources here
endif()
# ...
add_library(opentracing ...)
if(weak_symbol)
  target_compile_definitions(opentracing PUBLIC "ALLOW_WEAK_SYMBOL=1")
endif()
```

**Comment by rnburn** [comment]
@isaachier - I added in that test for weak symbols.

File: include/opentracing/dynamic_load.h [hunk 1]
```diff
@@ -4,10 +4,11 @@
 #include <opentracing/tracer.h>
 #include <opentracing/tracer_factory.h>
 #include <opentracing/version.h>
+#include <opentracing/config.h>
 #include <system_error>
 
-// OpenTracingMakeTracerFactory provides common hook that can be used to
-// create a TracerFactory from a dynamically loaded library. Users should prefer
+// OpenTracingMakeTracerFactory provides a common hook that can be used to
+// create an TracerFactory from a dynamically loaded library. Users should prefer
```
Comment (isaachier) [inline include/opentracing/dynamic_load.h]: Out of curiosity, why?

File: include/opentracing/dynamic_load.h [hunk 2]
```diff
@@ -35,11 +36,7 @@
 //      std::error_code error{rcode, *error_category};
 //   }
 extern "C" {
-#ifdef _MSC_VER
-__declspec(selectany) int (*OpenTracingMakeTracerFactory)(
-    const char* opentracing_version, const void** error_category,
-    void** tracer_factory);
-#else
+#ifdef OPENTRACING_BUILD_DYNAMIC_LOADING
```
Comment (isaachier) [inline include/opentracing/dynamic_load.h]: Thank you!

File: cmake/weak_symbol.cpp [hunk 3]
```diff
@@ -0,0 +1,3 @@
+void __attribute((weak)) f();
```
Comment (isaachier) [inline cmake/weak_symbol.cpp]: Is this the way to write it for all compilers? If not, I can devise a simple CMake loop to determine which, if any, weak symbol attribute notation works.

File: include/opentracing/dynamic_load.h (hunk 1 above)
Comment (rnburn) [reply include/opentracing/dynamic_load.h]: There are quite a few steps that need to be done if you want to call the OpenTracingMakeTracerFactory function directly, especially if you are doing it with `dlopen` and `dlsym`; furthermore, it's platform-dependent so they would be different on Linux and Windows.

DynamicallyLoadTracingLibrary lets you avoid the manual steps and will work on Windows too (assuming we add support at some point).

File: cmake/weak_symbol.cpp (hunk 3 above)
Comment (rnburn) [reply cmake/weak_symbol.cpp]: GCC, Clang, [ICC](https://software.intel.com/en-us/articles/performance-tools-for-software-developers-gnu-compatibility-attribute-weak-and-internal-linkage), and XL C [http://www-01.ibm.com/support/docview.wss?uid=swg27044475&aid=7] all use the same weak symbol syntax.

MSVC has something different, but it's not quite the same as weak symbols and would have to be handled differently. Probably best to deal with that whenever we add other pieces for MSVC.

**Comment by rnburn** [comment]
@isaachier -- does this answer everything for you?

File: cmake/weak_symbol.cpp (hunk 3 above)
Comment (isaachier) [reply cmake/weak_symbol.cpp]: OK makes sense to me. Just like to leverage build tools for compatibility when it is straightforward to do so.

**Pull request merged**